### ISSUE TITLE
Please test IPv6 first, then IPv4

### ISSUE THREAD

**Issue by romanrm** [ssllabs/ssllabs-scan#658]
Current order is IPv4, then IPv6.

Testing IPv4 first leads to the issue of webmasters closing the test result window not waiting for the IPv6 test to finish, and not noticing potential issues or misconfiguration on the IPv6 side.

https://tools.ietf.org/html/rfc6555#section-4.1 mandates giving preference to IPv6 when connecting, so dual-stack users are more likely to access the IPv6 resource. As such, having that resource tested last, or in many cases not tested at all, is actively harmful to the test's mission.

**Comment by DXGLdotinfo** [comment]
Good point, although most IPv6 issues I have experienced have been at the client side; my ISP and my router are very flaky when it comes to IPv6 support, only recently providing a reasonably optimized route to my server.

Perhaps SSL Labs tests IPv4 first in case querying the AAAA record results in a fatal error due to a misconfigured DNS server?

**Comment by romanrm** [comment]
> Perhaps SSL Labs tests IPv4 first in case querying the AAAA record results in a fatal error due to a misconfigured DNS server?

It always tests IPv4 first, not in case of any IPv6 failure.

Or do you mean it should keep doing that just because "what if IPv6 fails"? As I said, users will access IPv6 first, so any error on that side better be caught by the test, and not blindly ignored by the webmaster.

**Comment by DXGLdotinfo** [comment]
I ran a test on a website infected with Cloudflare, and it was listed as follows:
IPv6, IPv4, IPv6, IPv4

**Comment by romanrm** [comment]
For me it always tests IPv4 first, including on CloudFlare websites: https://romanrm.hk/2018-10-17T045705Z-qualys.png (same for the dev.ssllabs.com version).

**Comment by AliceWonderMiscreations** [comment]
For me it depends. Some servers it does IPv6 first, others IPv4 first. Seems to be consistent with those servers, e.g. when IPv6 is tested first it always is tested first and reverse. Not sure why.

**Comment by DXGLdotinfo** [comment]
Perhaps there is race condition in the logic that gathers the IP records from the DNS servers?

**Comment by smarek** [comment]
@dxgldotorg could it be https://en.wikipedia.org/wiki/Round-robin_DNS

>With each DNS response, the IP address sequence in the list is permuted. Usually, IP clients initially attempt connections with the first address returned from a DNS query, so that on different connection attempts, clients would receive service from different providers, thus distributing the overall load among servers.
